### PR TITLE
sci-libs/pdal: fix gdal-3.7 build

### PR DESCRIPTION
**Pull request by t0b3**
Closes: https://bugs.gentoo.org/906102
Closes: https://github.com/gentoo/gentoo/pull/30998

### CODE REVIEW

**Comment by espinafre** [comment]
Upstream has a different approach for this issue:

https://github.com/PDAL/PDAL/commit/7b4b2d942bc0390d370d5b90f31daf77148005f6

TLDR, instead of marking lvalues `const`, upstream made a `const_cast` on the rvalues.

The issue in bugs.gentoo.org (https://bugs.gentoo.org/906102 ) was updated to reflect the "upstream way":

https://bugs.gentoo.org/attachment.cgi?id=861527

Also, strictly speaking this fix isn't related to gcc-13, as upstream said that they had the bug with gcc-11.3.0 (see discussion on upstream merge request: https://github.com/PDAL/PDAL/pull/4030 ).  What is related to gcc-13 is the fix on the header vendor/lazperf/Extractor.hpp for pdal-2.4.3 (see patch on original issue: https://bugs.gentoo.org/attachment.cgi?id=861529 ); this fix isn't necessary for 2.5.3.

**Comment by t0b3** [comment]
@espinafre so do we need to patch 2.5.3 at all for gcc-13 or can we cancel this PR?

**Comment by espinafre** [comment]
@t0b3 , we need this to patch 2.5.3 AND 2.4.3 because of gdal-7.0; the other, slightly larger patch that involves lszperf/Extractor.hpp is needed ONLY for pdal-2.4.3, still because of gcc-13

**Comment by espinafre** [comment]
@t0b3 , without this patch pdal 2.5.3 and earlier won't build on a system that has gdal-3.7

**Comment by juippis** [comment]
So is this PR good to know? Does something else need to be added here?

**Comment by espinafre** [comment]
@juippis , @t0b3 , I suggest patching pdal-2.5.3 like this:

```patch
diff -Nru PDAL-2.5.3-src-orig/pdal/Geometry.cpp PDAL-2.5.3-src/pdal/Geometry.cpp
--- PDAL-2.5.3-src-orig/pdal/Geometry.cpp       2023-05-11 01:06:05.333184285 -0300
+++ PDAL-2.5.3-src/pdal/Geometry.cpp    2023-05-11 01:07:02.588477179 -0300
@@ -153,7 +153,7 @@
 
 bool Geometry::srsValid() const
 {
-    OGRSpatialReference *srs = m_geom->getSpatialReference();
+    OGRSpatialReference *srs = const_cast<OGRSpatialReference *>(m_geom->getSpatialReference());
     return srs && srs->GetRoot();
 }
 
@@ -172,7 +172,7 @@
         return StatusWithReason(-2,
             "Geometry::transform() failed.  NULL target SRS.");
 
-    OGRSpatialReference *inSrs = m_geom->getSpatialReference();
+    OGRSpatialReference *inSrs = const_cast<OGRSpatialReference *>(m_geom->getSpatialReference());
     SrsTransform transform(*inSrs, OGRSpatialReference(out.getWKT().data()));
     if (m_geom->transform(transform.get()) != OGRERR_NONE)
         return StatusWithReason(-1, "Geometry::transform() failed.");
```

This is because how upstream fixed the issue. Also, this patch doesn't fix the gcc-13 build issue, but fixes building against gdal-3.7; thus, the patch filename should be something along the lines of "pdal-2.5.3-fix-gdal-3.7.patch".

**Comment by espinafre** [comment]
@juippis , @t0b3 , the patch that fixes building pdal-2.4.3 against gcc-13 (as well as building against gdal-3.7.0) is as below (it doesn't apply to pdal-2.5.3):

```patch
diff -Nru PDAL-2.4.3-src-orig/pdal/Geometry.cpp PDAL-2.4.3-src/pdal/Geometry.cpp
--- PDAL-2.4.3-src-orig/pdal/Geometry.cpp       2023-05-10 23:13:39.643502276 -0300
+++ PDAL-2.4.3-src/pdal/Geometry.cpp    2023-05-10 23:11:33.142620506 -0300
@@ -140,7 +140,7 @@
 
 bool Geometry::srsValid() const
 {
-    OGRSpatialReference *srs = m_geom->getSpatialReference();
+    OGRSpatialReference *srs = const_cast<OGRSpatialReference *>(m_geom->getSpatialReference());
     return srs && srs->GetRoot();
 }
 
@@ -159,7 +159,7 @@
         return StatusWithReason(-2,
             "Geometry::transform() failed.  NULL target SRS.");
 
-    OGRSpatialReference *inSrs = m_geom->getSpatialReference();
+    OGRSpatialReference *inSrs = const_cast<OGRSpatialReference *>(m_geom->getSpatialReference());
     SrsTransform transform(*inSrs, OGRSpatialReference(out.getWKT().data()));
     if (m_geom->transform(transform.get()) != OGRERR_NONE)
         return StatusWithReason(-1, "Geometry::transform() failed.");
diff -Nru PDAL-2.4.3-src-orig/vendor/lazperf/Extractor.hpp PDAL-2.4.3-src/vendor/lazperf/Extractor.hpp
--- PDAL-2.4.3-src-orig/vendor/lazperf/Extractor.hpp    2023-05-10 23:13:39.613501276 -0300
+++ PDAL-2.4.3-src/vendor/lazperf/Extractor.hpp 2023-05-10 23:15:41.644241402 -0300
@@ -34,6 +34,7 @@
 
 #pragma once
 
+#include <cstdint>
 #include <cstring>
 #include <vector>
 
```

**Comment by espinafre** [comment]
@t0b3 , @juippis , I suggest reading the discussion concerning building pdal against gdal-3.7 in the upstream merge request: 
https://github.com/PDAL/PDAL/commit/7b4b2d942bc0390d370d5b90f31daf77148005f6

**Comment by t0b3** [comment]
@espinafre so I learn there was confusion of gcc13 and gdal37. you are always welcome to contribute a PR :smile:

**Comment by t0b3** [comment]
> So is this PR good to know? Does something else need to be added here?

good to go now

**Comment by espinafre** [comment]
@t0b3 lgtm. Sorry for the confusion of two different issues, and thanks for your helpfulness.

File: sci-libs/pdal/files/pdal-2.4.3-fix-gdal37.patch [hunk 1]
```diff
@@ -0,0 +1,21 @@
+diff -Nru PDAL-2.5.3-src-orig/pdal/Geometry.cpp PDAL-2.5.3-src/pdal/Geometry.cpp
```
Comment (thesamesam) [inline sci-libs/pdal/files/pdal-2.4.3-fix-gdal37.patch:1]: Please include commit links for patches.

Comment (t0b3) [reply sci-libs/pdal/files/pdal-2.4.3-fix-gdal37.patch:1]: @espinafre can you add this information?

Comment (t0b3) [reply sci-libs/pdal/files/pdal-2.4.3-fix-gdal37.patch:1]: and take into account upstream may soon change their mind https://github.com/PDAL/PDAL/pull/4043

Comment (t0b3) [reply sci-libs/pdal/files/pdal-2.4.3-fix-gdal37.patch]: @thesamesam links are in now :+1:

**Comment by thesamesam** [comment]
Thank you!
